datacube/utils/__init__.py

```diff
@@ -238,7 +291,28 @@ def remove_implicit_resolver(cls, tag_to_remove):
 NoDatesSafeLoader.remove_implicit_resolver('tag:yaml.org,2002:timestamp')
 
 
-def read_documents(*paths):
+def without_lineage_sources(doc, inplace=False):
+    """ Replace lineage.source_datasets with {}
+    """
+    lineage = doc.get('lineage', None)
+    if lineage is None:
+        return doc
+
+    if 'source_datasets' not in lineage:
+        return doc
+
+    if inplace:
+        doc['lineage']['source_datasets'] = {}
```

Comment:
Unfortunately, @jeremyh made this (and the rest) of dataset documents configurable with the whole MetadataType thing. See https://github.com/opendatacube/datacube-core/blob/fbefaa5f95704160133cd45df4cd5a12e1ea88f9/datacube/index/default-metadata-types.yaml#L54 for where it is defined for the `eo` metadata type.

Fortunately, in the two places this function is used it does have access to the Product, which is of a particular MetadataType, so can lookup what offset `sources` will be at. Still a total pain though...

Reply:
Good point, I keep forgetting about this indirection. It does mean that `sources` can not be skipped until top-level product is resolved (auto-matched), not a problem in `update`, but we do need this in `add` as well. I'll have a look what that implies.